src/main/java/hudson/remoting/FileSystemJarCache.java

```diff
@@ -117,6 +125,32 @@ protected URL retrieve(Channel channel, long sum1, long sum2) throws IOException
         }
     }
 
+    /**
+     * Get file checksum calculating it or retrieving from cache.
+     */
+    private Checksum fileChecksum(File file) throws IOException {
+        // When callers all request the checksum of a large jar the calls to
+        // forURL will all fall through to this method since the first caller's
+        // calculation may take a while. Hence re-check the cache at the start.
+        String location = file.getCanonicalPath();
+        if (checksumsByPath.containsKey(location)) {
```

Comment:
🐛 `checksumsByPath ` is not correctly synchronized here. `contains()` and `get()` go as separate calls, and the item may be removed between these two calls